orttraining/orttraining/core/graph/training_op_defs.cc

```diff
@@ -2192,6 +2192,65 @@ Example 4:
       .TypeConstraint("T2", {"tensor(bool)"}, "Constrain 'training_mode' type to boolean tensor.")
       .TypeConstraint("T3", {"tensor(uint32)"}, "Constrain 'mask' type to bit-packed uint32 tensor.");
 
+  ONNX_CONTRIB_OPERATOR_SCHEMA(BiasSoftmaxDropout)
+      .SetDomain(kMSDomain)
+      .SinceVersion(1)
+      .SetDoc(
+          "dropout_output, mask, softmax_output = Dropout(Softmax(data + bias), ratio), "
+          "Intended to specialize the Add + Softmax + Dropout pattern commonly found in transformer models.")
+      .Attr("axis", "apply softmax to elements for dimensions axis or higher", AttributeProto::INT,
+            static_cast<int64_t>(1))
+      .Attr("is_inner_broadcast",
+            "true if broadcast bias across input for dimensions broadcast_axis to axis-1, "
+            "otherwise broadcast bias across input for dimensions 0 to broadcast_axis-1",
+            AttributeProto::INT)
+      .Attr("seed", "(Optional) Seed to the random generator, if not specified we will auto generate one.",
+            AttributeProto::INT, OPTIONAL_VALUE)
+      .AllowUncheckedAttributes()
+      .Input(0, "data", "The input data as Tensor.", "T")
+      .Input(1, "bias", "The bias (or mask) as Tensor.", "T")
+      .Input(2, "ratio",
+             "The ratio of random dropout, with value in [0, 1). If this input was not set, "
+             "or if it was set to 0, the output would be a simple copy of the input. "
+             "If it's non-zero, output will be a random dropout of the scaled input, which is typically "
+             "the case during training. It is an optional value, if not specified it will default to 0.5.",
+             "T1", OpSchema::Optional)
+      .Output(0, "dropout_output", "The dropout output.", "T")
+      .Output(1, "mask", "The output mask of dropout.", "tensor(bool)")
```

Comment:
>"mask", "The  [](http://example.com/codeflow?start=17&length=13)

so we would lost the memory saving bitmask brought to us? is there a plan to make it a bit mask too?

Reply:
> Just make sure, from the PR description, I assume this new fusion will not affect our bitmask dropout replacement, is that right?

It's not right. Fusing BiasSoftmaxDropout will replace BitmaskDropout to boolean Dropout. I don't figure out a way to support the BitmaskDropout in the softmax kernel when the element count for each batch cannot divided by 32, this means the some UNIT32 bitmask will cross different warps so that it's not easy to set the bitmask using warp APIs. But even switch back to boolean Dropout, the perf is still better than BiasSoftmax+BitmaskDropout (the perf comparison in the description is this case). So no plan for now to support bitmask as it's not an easy task.

Reply:
OK. This brings out the issue we talked previously; different fusion rules might bring different implications on compute latency or memory consumptions. And plus some recompute stuff, the best fusion strategy could be different on variant models. It's interesting to see such "conflict" in graph fusion.